lib/evaluate/type.h

```diff
@@ -123,15 +129,22 @@ class DynamicType {
   constexpr const semantics::ParamValue *charLength() const {
     return charLength_;
   }
-  constexpr bool isPolymorphic() const { return isPolymorphic_; }
 
   std::string AsFortran() const;
   std::string AsFortran(std::string &&charLenExpr) const;
   DynamicType ResultTypeForMultiply(const DynamicType &) const;
 
   bool IsAssumedLengthCharacter() const;
+  constexpr bool IsPolymorphic() const {
+    return category_ == TypeCategory::Derived && kind_ > 0;
+  }
   constexpr bool IsUnlimitedPolymorphic() const {
-    return isPolymorphic_ && derived_ == nullptr;
+    return category_ == TypeCategory::Derived && derived_ == nullptr &&
+        kind_ == 1;
```

Comment:
After reviewing Tim's PR #539, I am not sure the name of this function is accurate. It does not accept `Type(*)` as unlimited polymorphic, but the standard says in 7.3.2:

> An entity that is declared using the TYPE(*) type specifier is assumed-type and is an unlimited polymorphic entity.

Would it make sens to rename this function to `IsClassStar` and/or to change it to accept `Type(*)` ?

Reply:
No, it's better to follow the standard and return true for `TYPE(*)`.